_posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md

```diff
@@ -0,0 +1,145 @@
+2. **users**: Creates the users and groups necessary to run docker, and the jupyterhub service itself.
+3. **certbot**: Runs "certbot" to create a signed SSL certificate for the site <link>certbot</link>, and sets up a cron job that will automatically re-generate the certificate before it expires.
+4. **docker**: Installs docker on the virtual machine, and sets up our dockerized jupyterhub system run from a single command.
+5. **jupyterhub**: Adds a service configuration for jupyterhub that launches our dockerized jupyterhub system when the VM boots up, which lets start and stop it like any other service.
+
+### If at first you don't succeed... 
+
+Each step in our ansible playbook is designed to be re-run if the a play fails to complete for any reason. There can be a lot of different steps to configure, but we keep things organized by putting all of that configuration information into two files that do only one thing: set ansible variables.
+
+One set-up file holds public information, and gets checked into the public repository. The second file is added to the .gitignore file; it contains all of the passwords, SSL certificates, secret keys and other details that should be kept private.
+
+All of the settings are in one place, so changes are fast and easy.
+
+For example, I was able to demo a new prototype version of our system on the same (production) server that was running our old code -- by changing the username of the service, and the port number, I could safely deploy our ansible script, confident in the knowledge that the new service would not interfere with the old one.
+
+## You've got to keep 'em separated... 
+
+<image src="{{ site.urlimg }}/kevin_hartmann/tangled-wires.jpg" /> 
+
+Docker runs virtual machines. We already have a virtual machine to run Jupyterhub on: so why Dockerize the service?
+
+For us, using docker simplifies development and testing. By using Docker, I can separate many of the issues related to running the virtual machine (and it's associated networking, services, security concerns, and so on) from the configuration of the Jupyterhub system itself.
+
+And by using Docker, I and other developers can spin up a copy of Jupyterhub on our computer, make any changes we want to the look and feel or operating features that our Jupyterhub system should have, and then deploy it.
+
+Within the Jupyterhub system, I further isolated the section of the python module that sets up all the web related settings for Jupyterhub (the css modules, the template files, and so on), by providing a symbolic link from the service directory to the relavent section of the python library module. That way, we can isolate the changes we make to the Jupyterhub front end from the python code that launches those changes; making it easy to change the web side of the system without worrying too much about the code itself.
+
+# What's this button do? 
+<image src="{{ site.urlimg }}/kevin_hartmann/cat.gif" /> 
```

Comment:
The cat gif is great but it tickles the eye a bit, can you get either a static image or one either smaller or less extreme in color transitions?